### PR TITLE
Implement selecting input device

### PR DESCRIPTION
**Pull request by cgfarmer4**
Allow for selection of Core Audio device to be used with `AVAudioEngine`. 

Please note this will only work on `macOS`.  Not sure theres a cleaner way to do this besides the `#if os(macOS)`.

https://streamable.com/vv0zjb

### CODE REVIEW

**Comment by cgfarmer4** [comment]
Seeing a test fail but looks unrelated to my changes?

```
Test Suite 'All tests' started at 2024-03-05 20:21:19.203.
Test Suite 'whisperkitPackageTests.xctest' started at 2024-03-05 20:21:19.205.
Test Suite 'FunctionalTests' started at 2024-03-05 20:21:19.205.
Test Case '-[WhisperKitTests.FunctionalTests testAsyncImplementation]' started.
WhisperKit/WhisperKit.swift:192: Fatal error: Unexpectedly found nil while unwrapping an Optional value
```

File: Examples/WhisperAX/WhisperAX/Views/ContentView.swift [hunk 1]
```diff
@@ -854,12 +886,27 @@ struct ContentView: View {
...
                     return
                 }
+                
+                #if os(macOS)
+                var deviceId: AudioDeviceID?
+                if self.selectedAudioInput != "No Audio Input",
+                   let devices = self.audioDevices,
+                   let device = devices.first(where: {$0.name == selectedAudioInput}) {
+                    deviceId = device.id
+                }
```
Comment (ZachNagengast) [inline Examples/WhisperAX/WhisperAX/Views/ContentView.swift:896]:
```suggestion
                }
                #endif
```

File: Examples/WhisperAX/WhisperAX/Views/ContentView.swift [hunk 2]
```diff
@@ -854,12 +886,27 @@ struct ContentView: View {
...
                         bufferEnergy = whisperKit?.audioProcessor.relativeEnergy ?? []
                     }
                 }
+                #else
+                try? audioProcessor.startRecordingLive() { _ in
+                    DispatchQueue.main.async {
+                        bufferEnergy = whisperKit?.audioProcessor.relativeEnergy ?? []
+                    }
+                }
+                #endif
```
Comment (ZachNagengast) [inline Examples/WhisperAX/WhisperAX/Views/ContentView.swift:909]:
```suggestion
```

**Review by ZachNagengast — CHANGES_REQUESTED** [review]
Very nice, thanks for the great demo video.

There may be a couple ways to clean up the various macos flags, it does seem to be a bit more than necessary:
1. Create a typealias for macos `typealias DeviceID = AudioDeviceID`
2. Use this to keep `startRecordingLive` a single interface for every OS
3. Then you can keep that inputDeviceId optional everywhere, and only have flags around the specific macos code that makes it non-nil.

Doing this should reduce the amount of repeated code inside the macos flags by a good amount, want to keep the forking logic per OS to a minimum (some is unavoidable)

As far as the UI - I'd like to see this selection inside the `controlsView`, potentially in between the reset and the settings buttons, although you'll need to make adjustments for the stream tab, which is slightly different than transcribe tab.

**Comment by cgfarmer4** [comment]
Awesome, thanks for the feedback. Ill work through this over the next couple days!

**Comment by cgfarmer4** [comment]
@ZachNagengast cleaned up the forks. Not that experienced in multi-platform so thanks for the suggestion. Less forks, cleaner code 💪🏻. Controls look like this now. Tried to figure out a way to do it without throwing off the balance and this seems to work ok. Happy to continue to iterate if you have more feedback.

image

Lastly, when I first plugged in my audio interface, I got a new exception from CoreAudio. `required condition is false: format.sampleRate == hwFormat.sampleRate`, rebooting my machine resolved this but not sure how much we care to guard against this here. Found some potential paths [here](https://stackoverflow.com/questions/66971504/i-got-crash-when-record-required-condition-is-false-format-samplerate-hwf) but they dont seem aligned with the need for `16k` output. Thoughts?

**Review by ZachNagengast — APPROVED** [review]
Looks great! Running the tests then good to merge once complete. For the transcribe tab in particular, I was thinking of having it centered between these buttons, but not a blocker - will check it out to see what works well before the next release.

**Comment by ZachNagengast** [comment]
Regarding the error, we do need to iron out requesting 16khz sample rate for the devices that support it, but that should be in it's own PR. We have a converter that will deal with anything not in 16khz but it adds a small & avoidable latency.

**Comment by cgfarmer4** [comment]
Nice! I can add it between the Transcribe buttons in a follow up PR, np.

**Pull request merged**